### PR TITLE
cmd/syncthing: Add CLI completion functionality (implements #8616)

### PR DESCRIPTION
**Pull request by danpadcz**
### Purpose

This implements CLI completion using the Kongplete module. As a side effect a CLI structure for syncthing/cli was created for kongplete to be able to parse and implement CLI completion.

This pull request implements issue #8616.

### Testing

I've tested the autocompletion manually, and it had worked, but I hadn't added any tests so as to test it automatically. Additionally, I ran `go run build.go test` with all tests passing.

The completion can be tested when running the output of `syncthing install-completions` in bash. Additionally, I would like to ask the maintainers, if and where should the install-completions command be placed?

## Authorship

Your name and email will be added automatically to the AUTHORS file
based on the commit metadata.

### CODE REVIEW

**Comment by danpadcz** [comment]
Hi, as was recommended in the original issue, I used kongplete for the cli completion, but since the `syncthing cli` subprogram uses "github.com/urfave/cli" instead of kong as the cli parser, I decided to create a struct representing the `syncthing cli` subprogram that I inserted into the kong cli parsing structure in `cmd/syncthing/main.go`. This enabled kongplete to create cli completion for the `syncthing cli` subprogram aswell. Is this a valid way of implementing cli completion for this project?

**Comment by AudriusButkevicius** [comment]
The whole point of the CLI is that it's auto generated from the structs, so we don't need to maintain the CLI, and every time we change the config, the CLI immediately supports setting that config element.

Maintaining this large struct by hand is definitely not the direction we want to go, as it will be out of date pretty quickly.
We should use the same reflective approach for generating completion as well.

**Comment by danpadcz** [comment]
That's understandable. Do you have an idea on how to implement the completion since it can't fully be autogenerated by kongplete due to the CLI subprogram using a different argument parser than kong?

**Comment by AudriusButkevicius** [comment]
I really don't know how shell completion works, as in, I don't know what you need to provide to the shell and at what point in time, but whatever that is, we should do that the same way we do the CLI generation, reflectively on the fly.

Seems that the library we use suggests it already supports it (https://cli.urfave.org/v2/examples/bash-completions/). The recli package we use for reflectiveness might need making it work with newer version of urfave/cli and perhaps we'd get completion if we just upgraded.

**Comment by AudriusButkevicius** [comment]
Actually, even v1 supports completion, so perhaps it's just switching the flag on.

**Comment by AudriusButkevicius** [comment]
Ok, the problem is a bit more convoluted, as the project actually uses two different command line parsers already.
One for normal entrypoint, the other one for CLI.
The normal entrypoint delegates to CLI after the CLI subcommand, so the completion for both (or even either maybe?) might be hard.

**Comment by AudriusButkevicius** [comment]
Seems that you enable that flag in urfave/cli it already does completion:
```
root@mail ~/syncthing # ./bin/syncthing cli config gui --generate-bash-completion
enabled
raw-address
raw-unix-socket-permissions
user
password
auth-mode
raw-use-tls
apikey
insecure-admin-access
theme
debugging
insecure-skip-host-check
insecure-allow-frame-loading
send-basic-auth-prompt
dump-json
help
h
```

So I think you just need a bit of glue (potentially fork of kongcomplete) to marry the two together.

**Comment by danpadcz** [comment]
In the end I tried to force urfave/cli completion generation to cooperate with the way kongplete does completion, but for the completion in the cli subcommand to work properly, we need to install it differently than kongplete does. Thats because urfave/cli returns all subcommands regardless of current partial input when doing completion, so I simply added a filter that removes all non-matching offered results.

So the changed installation command will be `complete -C path_to_syncthing_binary -X '!&*' syncthing`. 
This seems like a much better solution than having a manually created structure to me

**Comment by AudriusButkevicius** [comment]
Could you please rebase it on latest master. I'll try to give it a go at home tonight.

File: cmd/syncthing/main.go [hunk 1]
```diff
@@ -216,7 +217,11 @@ func main() {
 	// The "cli" subcommand uses a different command line parser, and e.g. help
 	// gets mangled when integrating it as a subcommand -> detect it here at the
 	// beginning.
-	if len(os.Args) > 1 && os.Args[1] == "cli" {
+	doCompgen := (strings.Contains(os.Getenv("COMP_LINE"), " cli") && os.Getenv("SHELL") == "/bin/bash")
```
Comment (AudriusButkevicius) [inline cmd/syncthing/main.go:220]: Some explanation on what the intent of these checks would be nice. Why do we care if its bash or sh?

**Comment by AudriusButkevicius** [comment]
Didn't realise quite a large PR landed that affects this area, so rebasing might be a bit of work. Sorry.

**Comment by danpadcz** [comment]
The changes on main today replaced urfave/cli with kong for all parts of the cli subprogram except for `cli config` and so completion can now be added for a large part of the project with one line of code :D 
The rewrites made also mean that my gluing the two parsers together no longer works, so I removed that. And since completion is missing for only one part, I figured it might be better to let it be for now?

File: go.mod [hunk 2]
```diff
@@ -57,19 +57,25 @@ require (
 	google.golang.org/protobuf v1.31.0
 )
 
+require github.com/willabides/kongplete v0.3.0
+
```
Comment (AudriusButkevicius) [inline go.mod:61]: Did you add it using `go get`?  The fact that it's in a separate block is strange.

**Review by AudriusButkevicius — APPROVED** [review]

File: go.mod (hunk 2 above)
Comment (danpadcz) [reply go.mod:61]: I think I did, should I move it into the first block instead?

**Comment by AudriusButkevicius** [comment]
I don't see us adding a command for installing the completions, so I've built this locally, but not obvious to me how to set it up, as I assumed you need `kongplete.InstallCompletions` command somewhere.

**Comment by danpadcz** [comment]
The `kongplete.InstallCompletions` command gives you a command you run in your shell that adds completion for the current shell. So for bash the command is for example of the format `complete -C /usr/bin/syncthing syncthing`.
A user that wants to install completions for all future shell instances will save this command in a script in `/etc/bash_completion.d/` or anywhere else if they prefer. 
For easier installation it could be nice to have the `kongplete.InstallCompletions`, somewhere, but I also am not sure where it would be intuitive to place and also the user will still need to know where to place the command according to their own shell preference.

**Comment by calmh** [comment]
I added an install-completions entrypoint, because otherwise I couldn't figure out how to test it on my zsh :)

We'll have to add a doc entry for how to use this.

**Pull request merged**